IQM Server Helper (#468)
NVIDIA/cuda-quantum: docs/sphinx/using/hardware.rst

the ``shots_count`` argument to ``cudaq.sample`` or ``cudaq.observe``. By default,
the ``shots_count`` is set to 1000.

.. code:: python

    cudaq.sample(kernel, shots_count=10000)

To see a complete example for using IonQ's backends, take a look at our :ref:`Python examples<python-examples-for-hardware-providers>`.

IQM
==================================

Setting Credentials
`````````````````````````

> Programmers of CUDA Quantum may access the IQM Server from either C++ or Python. Install `iqm-cortex-cli`,
> follow the `quick start guide <https://iqm-finland.github.io/cortex-cli/readme.html#using-cortex-cli>`__ to login and initialize the tokens file.

Reviewer: ```suggestion
Programmers of CUDA Quantum may access the IQM Server from either C++ or Python. Following the `quick start guide <https://iqm-finland.github.io/cortex-cli/readme.html#using-cortex-cli>`__, install `iqm-cortex-cli` and login to initialize the tokens file.
```

The path to the tokens file can either be passed explicitly via an environment variable or it will be loaded automatically if located in
the default location `~/.cache/iqm-cortex-cli/tokens.json`.

.. code:: bash

  export IQM_TOKENS_FILE="path/to/tokens.json"

Submission from C++
`````````````````````````

To target quantum kernel code for execution on an IQM Server,

> pass the flag ``--target iqm`` to the ``nvq++`` compiler. There is one mandatory string argument `--iqm-qpu-architecture`.
> The provided architecture must match the device architecture the compiled programs is being executed against.
> The hardware architecture for a specific IQM Server can be checked  via `https://<IQM server>/cocos/quantum-architecture`.

Reviewer: ```suggestion
pass the ``--target iqm`` flag to the ``nvq++`` compiler, along with a specified ``--iqm-qpu-architecture``. 

.. note:: 
    The ``--iqm-qpu-architecture`` is  a mandatory argument. This provided architecture must match 
    the device architecture that the program has been compiled against. The hardware architecture for a 
    specific IQM Server may be checked  via `https://<IQM server>/cocos/quantum- architecture`.
```

.. code:: bash

    nvq++ --target iqm --iqm-qpu-architecture Adonis src.cpp

Once the binary for a specific IQM QPU architecture is compiled, it can be executed against any IQM Server with the same QPU architecture:

.. code:: bash

    nvq++ --target iqm --iqm-qpu-architecture Adonis src.cpp -o program
    IQM_SERVER_URL=https://<Adonis IQM Server>/cocos program

>     # executing the same program against an IQM Server with a different underlying QPU architecture will result in an error

Reviewer: ```suggestion
    # Executing the same program against an IQM Server with a different underlying QPU 
    # architecture will result in an error.
```

    IQM_SERVER_URL=https://<Apollo IQM Server>/cocos program

To emulate the IQM machine locally, without submitting to the IQM Server,
you can also pass the ``--emulate`` flag to ``nvq++``. This will emit any target
specific compiler diagnostics, before running a noise free emulation.

.. code:: bash

>     nvq++ --emulate --target iq --iqm-qpu-architecture Adonis src.cpp

Reviewer 2: Typo, `iq` -> `iqm`

Author: Updated.
